// bcw/templates/Node-Server-Auth/server/controllers/AuthController.js
import express from "express";
import userService from "../services/UserService";
import { Authorize } from "../middleware/authorize";
import authService from "../services/AuthService";

//PUBLIC
export default class AuthController {
  constructor() {
    this.router = express
      .Router()
      .post("/register", this.register)
      .post("/login", this.login)
      .use(Authorize.authenticated)
      .get("/authenticate", this.authenticate)
      .delete("/logout", this.logout);
  }

  async register(req, res, next) {
    try {
      let user = await authService.register(req.body);
      req.session.uid = user._id; //SET THE SESSION UID (SHORT FOR USERID)
      res.status(201).send(user);
    } catch (err) {
      next(err);
    }
  }

  async login(req, res, next) {
    try {
      let user = await authService.login(req.body);
      res.send(user);
    } catch (err) {
      next(err);
    }
  }

  async authenticate(req, res, next) {
    try {
      let user = await authService.authenticate(req.session.uid);
      res.send(user);
    } catch (err) {
      next(err);
    }
  }

  async logout(req, res, next) {
    try {
      req.session.destroy(err => {
        if (err) {
          return res.send(err);
        }
        return res.send({
          message: "Logout Successful"
        });
      });
    } catch (error) {
      next(error);
    }
  }
}
